src/auth/policy/resolver.go

```diff
@@ -1,41 +1,349 @@
 package policy
 
 import (
+	"fmt"
+	"strings"
+
+	"github.com/armon/go-radix"
 	"github.com/consensys/quorum-key-manager/src/auth/types"
 )
 
+var (
+	wildCardSuffix = "*"
+)
+
 type Operation struct {
 	Action string
 	Path   string
 }
 
-type Result struct {
-	isAllowed bool
-	err       error
+type Result interface {
+	Allowed() bool
+	Error() error
+}
+
+// Resolver is responsible to control whether an operation is authorized or not
+// depending on the set of policies attached to the resolver
+type Resolver interface {
+	IsAuthorized(op *Operation) Result
 }
 
-func (res *Result) Allowed() bool {
-	return res.isAllowed
+type statement struct {
+	types.Statement
+	policy *types.Policy
 }
 
-func (res *Result) Error() error {
-	return res.err
+type result struct {
```

Comment:
Why did you need to store all this information as part of the result object. Isn't it enough to return a true or false as the result of resolving the access requested Operation ?

Reply:
To return extra information details about why an operation was rejected to a user. It helps users configure/debug their policies.

Typically when having a large number of policies, it would be almost impossible to properly configure/debug the policies.

Btw this is what AWS does.